[End- to End Test code sprint] Add SEVIRI METEOSAT-8 and METEOSAT-11 to end-to-end testing

azadeh-gh opened the pull request:
(#765)

End-to-end testing for Spinning Enhanced Visible and InfraRed Imager (**SEVIRI**) from METEOSAT-8 and to METEOSAT-11

**New files include:**
- `parm/ioda/bufr2ioda//bufr2ioda_sevcsr.json`  : JSON containing data format, sensor/bufr2ioda_sevcsr.json, and satellite information
-  `ush/ioda/bufr2ioda/bufr2ioda_sevcsr.py` : bufr2ioda python code for converting SEVIRI METEOSAT-8 and METEOSAT-11 from BUFR to IODA 
- `parm/atm/obs/config/seviri_m08.yaml.j2`
- `parm/atm/obs/config/seviri_m11.yaml.j2`
- `parm/atm/obs/lists/gdas_prototype_3d.yaml.j2`

RussTreadon-NOAA committed:
update prototype_3d observation list to process satwind_goes-16, update obsfile for satwnd.abi_goes-16

RussTreadon-NOAA committed:
Merge develop at 957a2fc into feature/gdas-validation

RussTreadon-NOAA committed:
update GDAS/orion.lua to use spack-stack-1.5.1, bufr/12.0.1, and atlas/0.35.0

emilyhcliu committed:
Remove a double-declaired "obs post filters" in satwnd GOES YAMLs.

CoryMartin-NOAA committed:
Merge branch 'develop' into feature/gdas-validation

emilyhcliu committed:
Modify BUFR converter and filter yamls for GOES AMV (#724)

Two updates for GOES AMV

1. BUFR Converter:
- remove wind direction and speed
- change sensorZenithAngle to satelliteZenithAngle
2. UFO Filters:
- change `sensorZenithAngle` to `satelliteZenithAngle`
- add `linear obs operator` section
The declaration of `linear obs operator" is necessary for GOES AMV since
there is no linearized component for applying the 10-meter factor.

The increments look reasonable:
```
0: ----------------------------------------------------------------------------------------------------
  0: Increment print | number of fields = 8 | cube sphere face size: C768
  0: eastward_wind                                | Min:-1.642335e-01 Max:+1.660654e-01 RMS:+2.746685e-04
  0: northward_wind                               | Min:-2.120068e-01 Max:+1.663539e-01 RMS:+2.631295e-04
  0: air_temperature                              | Min:+0.000000e+00 Max:+0.000000e+00 RMS:+0.000000e+00
  0: surface_pressure                             | Min:+0.000000e+00 Max:+0.000000e+00 RMS:+0.000000e+00
  0: specific_humidity                            | Min:+0.000000e+00 Max:+0.000000e+00 RMS:+0.000000e+00
  0: cloud_liquid_ice                             | Min:+0.000000e+00 Max:+0.000000e+00 RMS:+0.000000e+00
  0: cloud_liquid_water                           | Min:+0.000000e+00 Max:+0.000000e+00 RMS:+0.000000e+00
  0: ozone_mass_mixing_ratio                      | Min:+0.000000e+00 Max:+0.000000e+00 RMS:+0.000000e+00
  0: ----------------------------------------------------------------------------------------------------
```

The UFO filter results vs. GSI  -- looks good


![ufo_abi_goes-16_omf_windEastward_qc_time1](https://github.com/NOAA-EMC/GDASApp/assets/36091766/017aeabc-d6ca-4643-b037-16f46dde76a0)


![gsi_abi_goes-16_omf_windEastward_qc_time1](https://github.com/NOAA-EMC/GDASApp/assets/36091766/d99599a2-3f36-4ef6-aa2e-4611500b8c43)

CoryMartin-NOAA committed:
Make the python bufr2ioda driver more generic (#725)

Rather than each new ob type being manually added to the list, we can
use `glob.glob` to get the list of available observation types.

CoryMartin-NOAA committed:
Merge branch 'develop' into feature/gdas-validation

emilyhcliu committed:
Modify BUFR converter and filter yamls for MetoOp Scatwind (#731)

Two updates for MetOp SCATWIND; one minor updates for GOES AMV 

1. BUFR Converter:
    - remove wind direction and speed
    - add `stationElevation` as zero 

2. UFO Filters:
    - add linear obs operator section
The declaration of `linear obs operator" is necessary for GOES AMV since
there is no
      linearized component for applying the 10-meter factor.
- add `Gaussian Thinning` filter to thin data with 75 km box (equal box
size, prefer data close to central of the box)

3. For GOES AMV YAML files, change output data extension from `nc4` to
`nc`

UFO vs. GSI O-F and QC comparison: O-F and data count passed look close
![ufo_ascat_metop-b_omf_windEastward_qc
(1)](https://github.com/NOAA-EMC/GDASApp/assets/36091766/f234543e-e0b5-4c6e-add9-3c7365c0c469)

![gsi_ascat_metop-b_omf_windEastward_qc](https://github.com/NOAA-EMC/GDASApp/assets/36091766/37f4fba8-bd2a-4a54-b45e-4055e3ef3142)

3Dvar increments are reasonable with MetOp-B SCATWIND alone.
```
0: ----------------------------------------------------------------------------------------------------
  0: Increment print | number of fields = 8 | cube sphere face size: C768
  0: eastward_wind                                | Min:-1.741092e+00 Max:+1.712287e+00 RMS:+2.282584e-03
  0: northward_wind                               | Min:-1.825957e+00 Max:+1.930385e+00 RMS:+2.496075e-03
  0: air_temperature                              | Min:+0.000000e+00 Max:+0.000000e+00 RMS:+0.000000e+00
  0: surface_pressure                             | Min:+0.000000e+00 Max:+0.000000e+00 RMS:+0.000000e+00
  0: specific_humidity                            | Min:+0.000000e+00 Max:+0.000000e+00 RMS:+0.000000e+00
  0: cloud_liquid_ice                             | Min:+0.000000e+00 Max:+0.000000e+00 RMS:+0.000000e+00
  0: cloud_liquid_water                           | Min:+0.000000e+00 Max:+0.000000e+00 RMS:+0.000000e+00
  0: ozone_mass_mixing_ratio                      | Min:+0.000000e+00 Max:+0.000000e+00 RMS:+0.000000e+00
  0: ----------------------------------------------------------------------------------------------------

```

3DVar increments are also reasonable with MetOp-B + MetOp-A assimilated

CoryMartin-NOAA committed:
Have the driver script process both python scripts and YAMLs + bufr2ioda.x (#730)

This PR allows for all obtypes to be processed in two loops.

We still need to figure out a way to process these in parallel
(somewhat).

CoryMartin-NOAA committed:
Use parallel processing to speed up obs processing (#733)

Using python multiprocessing to generate obs in parallel. The current
list of obs goes from 10+ minutes to completing in ~5.5 minutes.

CoryMartin-NOAA committed:
Fix YAML so minimization works for satwind G17 (#743)

@RussTreadon-NOAA noted that `obs linear operator` should be `linear obs
operator` in the GOES-17 satwind YAML. This fixes that.

CoryMartin-NOAA committed:
Consolidate parallel processing (#745)

This reduces runtime further to run everything in one parallel pool.
Now, things complete in under 4 minutes (as of now).

BrettHoover-NOAA committed:
Add AHI/Himawari-8 assimilation to end-to-end GDASApp validation (#746)

Adding satwinds from the Advanced Himawari Imager (AHI) from Himawari-8
to GDASApp end-to-end testing

new files include:
parm/atm/obs/config/satwind_ahi_h8.yaml: QC filter YAML for AHI
Himawari-8 satwinds
parm/ioda/bufr2ioda/bufr2ioda_satwind_amv_ahi.json: JSON containing data
format, sensor, and satellite information for AHI Himawari-8 satwinds
ush/ioda/bufr2ioda/bufr2ioda_satwind_amv_ahi.py: bufr2ioda code for
extracting AHI Himawari-8 satwinds from BUFR

modified files include:
ush/ioda/bufr2ioda/run_bufr2ioda.py: added `"satwind_amv_ahi"` to list
`BUFR_py`

See #741 for details on testing. JEDI/GSI comparisons look good with GSI
thinning turned off, but there are large ob-count disparities both
before and after QC when comparing GSI+thinning to JEDI+thinning. The
tested `Gaussian Thinning` filter is included in the YAML but commented
out with a note.

---------

Co-authored-by: Brett Hoover <bhoover@Orion-login-1.HPC.MsState.Edu>
Co-authored-by: Cory Martin <cory.r.martin@noaa.gov>

CoryMartin-NOAA committed:
Merge branch 'develop' into feature/gdas-validation

CoryMartin-NOAA committed:
Write the bufr2ioda config files in DATA (#748)

Instead of writing the filled out `.json` and `.yaml` files to
`$COM_OBS`, we write them to `$DATA` so that they get purged unless
`KEEPDATA="YES"`

XuanliLi-NOAA committed:
Update GNSSRO assimilation to end-to-end GDASApp validation (#753)

Updating GNSSRO converter and adding the yaml file to GDASApp end-to-end
testing based on [#712](https://github.com/NOAA-EMC/GDASApp/pull/712).

New file:
parm/atm/obs/config/gnssro.yaml: Added QC filters in the YAML

Updated files:
parm/ioda/bufr2ioda/bufr2ioda_gnssro_bufr.json: Updated the satellite
information
ush/ioda/bufr2ioda/bufr2ioda_gnssro_bufr.py: Updated pccf, qfro, and
satelliteAscending flag

Refer to [issue #750](https://github.com/NOAA-EMC/GDASApp/issues/750)
for testing details. JEDI/GSI comparisons revealed a cutoff near
obs=0.03 Rad due to super refraction 2. The number of observations after
QC in JEDI is also inconsistent with GSI.

emilyhcliu committed:
Update for ATMS for End-to-End Test (#757)

List of items to be added in this PR:
- BUFR Converter: Add YAML for converting ATMS BUFR to IODA
   - The ObsValue is antenna temperature
   - Input BUFR is normal feel 

- Testing YAML: re-evaluation results due to update CRTM from v2.3.0 to
v2.4.1-jedi.1
   - atms_npp_noqc.yaml --- 100% replication
   - atms_npp.yaml  --- 100 % replication
   - atms_n20_noqc.yaml --- 100 % replication
- atms_n20.yaml --- ~ 1100% replication: two channel 7 observations
difference
  
- Config YAML: 
* add two read routine QC: data thinning and removal of data from scan
edge
   - atms_npp.yaml
   - atms_n20.yaml   

- Paring PR in JCSDA UFO repository
- This PR also has a [paring PR in JCSDA UFO
repository](https://github.com/JCSDA-internal/ufo/pull/3094).
   - It is OK to merge this one into GDASApp before the paring UFO PR
- The related test data (obs and geoval files from GSI) for UFO
Evaluation have been updated (due to CRTM-2.4.1 update) on ORION and
HERA.


Notes:
- End-to-end testing completed without problem. 
- However, the filter results show one of the filters, 88-165 GHz
scattering index check, tossed more data compared to GSI.
- The excess screening from the 88-165 GHz scattering index check is
under investigation and will be reported separately.
- Please check comments below for validation results

BrettHoover-NOAA committed:
[End-to-End Test Code Sprint] Add SEVIRI METEOSAT-8 satwinds to end-to-end testing (#764)

Adding satwinds from the Spinning Enhanced Visible and InfraRed Imager
(SEVIRI) from METEOSAT-8 to GDASApp end-to-end testing

new files include:
parm/atm/obs/config/satwind_seviri_m8.yaml: QC filter YAML for SEVIRI
METEOSAT-8 satwinds
parm/ioda/bufr2ioda/bufr2ioda_satwind_amv_seviri.json: JSON containing
data format, sensor, and satellite information for SEVIRI METEOSAT-8
satwinds
ush/ioda/bufr2ioda/bufr2ioda_satwind_amv_seviri.py: bufr2ioda code for
extracting SEVIRI METEOSAT-8 satwinds from BUFR

End-to-end testing results are provided in #758 - Note: Thinning filter
is provided in YAML but turned off due to irreconcilable differences
with GSI in testing.

---------

Co-authored-by: Brett Hoover <bhoover@Orion-login-1.HPC.MsState.Edu>

Commit:
[End- to End Test code sprint] Add SEVIRI METEOSAT-8 and METEOSAT-11 (#765)
End-to-end testing for Spinning Enhanced Visible and InfraRed Imager (SEVIRI) from METEOSAT-8 and to METEOSAT-11
new files include:
parm/ioda/bufr2ioda//bufr2ioda_sevcsr.json : JSON containing data format, sensor/bufr2ioda_sevcsr.json, and satellite information
ush/ioda/bufr2ioda/bufr2ioda_sevcsr.py: bufr2ioda code for extracting SEVIRI METEOSAT-8 and METEOSAT-11 from BUFR

RussTreadon-NOAA commented:
**Orion test**
Install 
- `parm/ioda/bufr2ioda/bufr2ioda_sevcsr.json`
- `ush/ioda/bufr2ioda/bufr2ioda_sevcsr.py`

in a working copy of `feature/gdas-validation`.   

Copy `seviri_m08.yaml` and `seviri_m11.yaml` from `parm/atm/obs/testing` to `/parm/atm/obs/config`.  Update both for use in g-w.  It's probable that the updated seviri yamls require additional changes for thinning, qc, etc.   This is just a starting point to allow `fv3jedi_var.x` execution.

Ensure seviri satbias, satbias_cov, and tlapse files are in appropriate `COMROT` directory.   Add `seviri_m08` and `seviri_m11` to `parm/atm/obs/lists/gdas_prototype_3d.yaml`

Execute prepatmiodaobs, atmanlinit, and atmanlrun.  The first two jobs run to completion.  SEVIRI dump files are written to the appropriate COM_OBS directory
```
-rw-r--r-- 1 rtreadon da   6353993 Nov 22 11:07 gdas.t00z.seviri_m11.tm00.nc
-rw-r--r-- 1 rtreadon da   4862394 Nov 22 11:07 gdas.t00z.seviri_m08.tm00.nc
```

atmanlrun begins but `fv3jedi_var.x` fails with the exception
```
313: Exception:         Reason: An exception occurred inside ioda while opening a variable.
313:    name:   MetaData/sensorAzimuthAngle
313:    source_column:  0
313:    source_filename:        /work2/noaa/da/rtreadon/gdas-validation/global-workflow/sorc/gdas.cd/ioda/src/engines/ioda/src/ioda/Has_Variables.cpp
313:    source_function:        ioda::Variable ioda::detail::Has_Variables_Base::open(const std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>> &) const
313:    source_line:    108
313:
313: Exception: oops::Variational<FV3JEDI, UFO and IODA observations> terminating...
```
Notice that atms ioda format dump files contain MetaData `float sensorAzimuthAngle(Location)`.   This MetaData field is missing from seviri ioda format dump files.  _Should it be present?_

BrettHoover-NOAA commented:
I changed `sensorZenithAngle` to `satelliteZenithAngle` in my AMV code to match changes that @emilyhcliu had made - its possible that there could be a variable name mismatch between different parts of the code-base as well

BrettHoover-NOAA committed:
Added SEVIRI/METEOSAT-11 satwinds to end-to-end validation (#767)

Small code-change to include SEVIRI/METEOSAT-11 satwinds in end-to-end
validation. This uses the hook-ins already provided by the
SEVIRI/METEOSAT-8 satwinds code that was merged in
https://github.com/NOAA-EMC/GDASApp/pull/764.

New file: parm/atm/obs/config/satwind_seviri_m11.yaml - this is nearly a
direct copy of parm/atm/obs/config/satwind_seviri_m8.yaml, since they
utilize the same filters. Only the header information has changed.

Modified file: parm/ioda/bufr2ioda/bufr2ioda_satwind_amv_seviri.json -
METEOSAT-11 was added to the `satellite_info` group

End-to-end testing looks very similar to what was seen for
SEVIRI/METEOSAT-8 satwinds, detailed in
https://github.com/NOAA-EMC/GDASApp/issues/758. A brief run-down:

(LW)IR winds are in QC agreement down to a 0.02% difference
(cloud-top)WV winds are in QC agreement down to a 0.20% difference
VIS winds differ by 4.5%, but there are only 22 accepted winds in GSI
and 23 in JEDI, the difference is a single satwind

All ob, HofX, and ob-minus-HofX differences look good

Increment summary, from gdasatmanlrun.log:
```
  0: ----------------------------------------------------------------------------------------------------
  0: Increment print | number of fields = 8 | cube sphere face size: C768
  0: eastward_wind                                | Min:-3.479765e-01 Max:+3.430433e-01 RMS:+4.990993e-04
  0: northward_wind                               | Min:-4.226316e-01 Max:+3.376561e-01 RMS:+4.889429e-04
  0: air_temperature                              | Min:+0.000000e+00 Max:+0.000000e+00 RMS:+0.000000e+00
  0: surface_pressure                             | Min:+0.000000e+00 Max:+0.000000e+00 RMS:+0.000000e+00
  0: specific_humidity                            | Min:+0.000000e+00 Max:+0.000000e+00 RMS:+0.000000e+00
  0: cloud_liquid_ice                             | Min:+0.000000e+00 Max:+1.618770e-20 RMS:+1.293217e-23
  0: cloud_liquid_water                           | Min:+0.000000e+00 Max:+1.474788e-19 RMS:+2.167418e-22
  0: ozone_mass_mixing_ratio                      | Min:+0.000000e+00 Max:+0.000000e+00 RMS:+0.000000e+00
  0: ----------------------------------------------------------------------------------------------------
```

Co-authored-by: Brett Hoover <bhoover@Orion-login-1.HPC.MsState.Edu>

emilyhcliu committed:
A few updates for ATMS end-to-end testing (#768)

In [ previous PR](https://github.com/NOAA-EMC/GDASApp/pull/757) for
ATMS, the end-to-end testing was done without data thinning. And the
excessive QC filtering in clear-sky areas, compared to GSI, was found.
(See
[plots](https://github.com/NOAA-EMC/GDASApp/pull/757#issuecomment-1819753244))

This PR includes two fixes for ATMS in the end-to-end testing and one
new feature:
1. fix the excess QC filtering in clear-sky areas. ---> This has a
paring [UFO PR #3122](https://github.com/JCSDA-internal/ufo/pull/3122)
2. add data thinning 
3. add the diagnostic flags (QC) --- **This can [reproduce QC
flags](https://github.com/NOAA-EMC/GDASApp/pull/768#issuecomment-1828209921)
from GSI in UFO**

**To test updates in this PR, please check out the UFO branch:
[feature/satrad](https://github.com/JCSDA-internal/ufo/tree/feature/satrad)
from JCSDA-internal in gdas-validation. This branch consolidates all
proposed code changes to UFO for gdas validation:
UFO PR #https://github.com/JCSDA-internal/ufo/pull/3122
UFO PR #https://github.com/JCSDA-internal/ufo/pull/3121
UFO PR #https://github.com/JCSDA-internal/ufo/pull/3094

---------

Co-authored-by: Cory Martin <cory.r.martin@noaa.gov>

Commit:
Updated bufr2ioda_sevcsr.py to add sensorAzimuthAngle
Added seviri_m08 and seviri_m11 YAML files for config

Commit:
Add SEVIRIinto the data list for 3dvar

RussTreadon-NOAA commented:
Refresh working copy of gdas-validation with SEVIRI updates.   Run prepatmiodaobs, atmanlinit, and atmanlrun.   `fv3jedi_var.x` terminated with

```
  0: OOPS_TRACE[0] ufo::Variable(conf) done
  0: OOPS_TRACE[0] ufo::Variable(config) start LocalConfiguration[root={name => brightnessTemperature , channels => (nil)}]
326: Exception: The key '.' is set to neither a string nor a map with the correct keys.
326: Bad Conversion: Cannot convert (nil) (Nil) to std::string
326: Bad Conversion: Cannot convert {name => brightnessTemperature , channels => (nil)} (OrderedMap) to std::string  (/work2/noaa/da/rtreadon/gdas-validation/global-workflow/sorc/gdas.cd/ufo/src/ufo/utils/parameters/ParameterTraitsVariable.cc +70 get)
378: Exception: The key '.' is set to neither a string nor a map with the correct keys.
378: Bad Conversion: Cannot convert (nil) (Nil) to std::string
378: Bad Conversion: Cannot convert {name => brightnessTemperature , channels => (nil)} (OrderedMap) to std::string  (/work2/noaa/da/rtreadon/gdas-validation/global-workflow/sorc/gdas.cd/ufo/src/ufo/utils/parameters/ParameterTraitsVariable.cc +70 get)
380: Exception: The key '.' is set to neither a string nor a map with the correct keys.
380: Bad Conversion: Cannot convert (nil) (Nil) to std::string
380: Bad Conversion: Cannot convert {name => brightnessTemperature , channels => (nil)} (OrderedMap) to std::string  (/work2/noaa/da/rtreadon/gdas-validation/global-workflow/sorc/gdas.cd/ufo/src/ufo/utils/parameters/ParameterTraitsVariable.cc +70 get)
```
Issue with input yaml file or code?   

Run directory is `/work2/noaa/stmp/rtreadon/RUNDIRS/gdas_eval_satwind_JEDI/gdasatmanl_00_seviri`.   Log file is `/work2/noaa/da/rtreadon/gdas-validation/comrot/gdas_eval_satwind_JEDI/logs/2021080100/seviri_gdasatmanlrun.log`

CoryMartin-NOAA commented on parm/atm/obs/config/seviri_m11.yaml:
```diff
@@ -0,0 +1,328 @@
+- filter: RejectList
+  filter variables:
+  - name: brightnessTemperature
+    channels: 
```
This I think is the issue @RussTreadon-NOAA is finding in his test

RussTreadon-NOAA replied:
Good catch, @CoryMartin-NOAA .  I added `4-11` and reran atmanlinit & atmanlrun.   atmanlrun died with
```
 39: Exception: UserError: obs bias.covariance section missing from the YAML file
 41: Exception: UserError: obs bias.covariance section missing from the YAML file
  3: Exception: UserError: obs bias.covariance section missing from the YAML file

```
I know how to fix this.

RussTreadon-NOAA commented:
Next `fv3jedi_var.x` error is

```
  0: OOPS_TRACE[0] UnstructuredInterpolator::UnstructuredInterpolator start
346: Exception: Bad parameter: /: Mandatory parameter 'obserr_bound_max' not found  (/work2/noaa/da/rtreadon/gdas-validation/glo\
bal-workflow/sorc/gdas.cd/oops/src/oops/util/parameters/RequiredParameter.h +135 deserialize)
347: Exception: Bad parameter: /: Mandatory parameter 'obserr_bound_max' not found  (/work2/noaa/da/rtreadon/gdas-validation/glo\
bal-workflow/sorc/gdas.cd/oops/src/oops/util/parameters/RequiredParameter.h +135 deserialize)
348: Exception: Bad parameter: /: Mandatory parameter 'obserr_bound_max' not found  (/work2/noaa/da/rtreadon/gdas-validation/glo\
bal-workflow/sorc/gdas.cd/oops/src/oops/util/parameters/RequiredParameter.h +135 deserialize)
```

CoryMartin-NOAA commented on parm/atm/obs/config/seviri_m08.yaml:
```diff
@@ -0,0 +1,328 @@
+        channels: 4-11
+        options:
+          channels: 4-11
+      obserr_boundax: [ 2.0, 4.0, 3.5, 2.0, 2.0, 2.0, 2.0, 3.0 ]
```
@RussTreadon-NOAA here's your next issue, a typo. @azadeh-gh did you miss any commits when testing this that are not included in this PR?

azadeh-gh commented:
Thank you @RussTreadon-NOAA and @CoryMartin-NOAA for your tests/comments. Sorry I made a mistake, I wanted to push these commits to my [feature/Azadeh_gdas-validation](https://github.com/NOAA-EMC/GDASApp/tree/feature/Azadeh_gdas-validation) branch 
I updated the python code and added two Seviri YAMLS to config but my "gdasatmanlinit" run fails with "Permission denied" error.
 I get permission denied for all 20210801.000000 files in
 /work2/noaa/da/gholoubi/gdas-validation/comrot/gdas_eval_seviri2_JEDI/gdas.20210731/18/model_data/atmos/restart

Any suggestions?

CoryMartin-NOAA commented:
@azadeh-gh my guess is it is because you don't have `stmp` group access on Orion. Didn't we discuss this a couple of weeks ago (or was that someone else who also doesn't have it?)

RussTreadon-NOAA commented:
Next `fv3jedi_var.x` error after fixing `obserr_bound_max`

```
194: Exception:         Reason: An exception occurred inside ioda while opening a variable.
194:    name:   MetaData/solarAzimuthAngle
194:    source_column:  0
194:    source_filename:        /work2/noaa/da/rtreadon/gdas-validation/global-workflow/sorc/gdas.cd/ioda/src/engines/ioda/src/ioda/Has_Variables.cpp
194:    source_function:        ioda::Variable ioda::detail::Has_Variables_Base::open(const std::__cxx11::basic_string<char, std\
::char_traits<char>, std::allocator<char>> &) const
194:    source_line:    108
194:
194: Exception: oops::Variational<FV3JEDI, UFO and IODA observations> terminating...
```

azadeh-gh commented:
@CoryMartin-NOAA Yes, I still do not have stmp access (sent request to help desk), but I changed the DATAROOT from stmp to da, and can see other files in /work2/noaa/da/gholoubi/gdas-validation/comrot/gdas_eval_seviri2_JEDI/gdas.20210731/18/model_data/atmos/restart

CoryMartin-NOAA commented:
@azadeh-gh can you point me to your logfile so I can take a look at where the permission denied problem is occurring?

azadeh-gh commented:
@RussTreadon-NOAA solarAzimuthAngle and sensorAzimuthAngle are missing parameters in Seviri bufr file (bmiss in gsi read_seviri.f90).

based on obsmod.F90 :
def bmiss            - parameter to define missing value from bufr
! Set bufr missing value
#if def ibm_sp
  real(r_kind), parameter:: bmiss = 1.0e11_r_kind
#else
  real(r_kind), parameter:: bmiss = 1.0e9_r_kind
#endif

I used "fillval=float32_fill_value" for sensorAzimuthAngle is this correct? or I need to insert 1.0e11?

azadeh-gh commented:
@CoryMartin-NOAA  here is the log file:
/work2/noaa/da/gholoubi/gdas-validation/comrot/gdas_eval_seviri2_JEDI/logs/2021080100/gdasatmanlinit.log

CoryMartin-NOAA commented:
@azadeh-gh are you able to run the following commands?:
`cat /work2/noaa/da/gholoubi/gdas-validation/comrot/gdas_eval_seviri2_JEDI/gdas.20210731/18//model_data/atmos/restart/20210801.000000.coupler.res`
`cat /work2/noaa/da/cmartin/UFO_eval/data/para/output_ufo_eval_aug2021/2021080100/gdas.20210731/18/atmos/RESTART/20210801.000000.coupler.res` ?

RussTreadon-NOAA commented:
> @RussTreadon-NOAA solarAzimuthAngle and sensorAzimuthAngle are missing parameters in Seviri bufr file (bmiss in gsi read_seviri.f90).
> 
> based on obsmod.F90 : def bmiss - parameter to define missing value from bufr ! Set bufr missing value #if def ibm_sp real(r_kind), parameter:: bmiss = 1.0e11_r_kind #else real(r_kind), parameter:: bmiss = 1.0e9_r_kind #endif
> 
> I used "fillval=float32_fill_value" for sensorAzimuthAngle is this correct? or I need to insert 1.0e11?

Great question which prompts two additional questions
1. What do we encode (data type and range) for other satellite_sensors which have a valid solarAzimuthAngle?   
2. Does UFO/IODA have something akin to the GSI `bmiss`?  If so, what do UFO/IODA expect when a metadata field has a missing value?

azadeh-gh commented:
@CoryMartin-NOAA No, I get permission denied

CoryMartin-NOAA commented:
@azadeh-gh fascinating. I wonder why, you are in the `da` group...

azadeh-gh commented:
@CoryMartin-NOAA I can run :  cat /work2/noaa/da/gholoubi/gdas-validation/comrot/gdas_eval_seviri2_JEDI/gdas.20210731/18//model_data/atmos/restart/20210801.020000.coupler.res

I only get permission denied for 20210801.000000 files

CoryMartin-NOAA commented:
@azadeh-gh thanks I'll check and see if there are any differences

CoryMartin-NOAA commented:
@azadeh-gh ha!
```
-rw-r----- 1 cmartin stmp 300 Apr 25  2022 /work2/noaa/da/cmartin/UFO_eval/data/para/output_ufo_eval_aug2021/2021080100/gdas.20210731/18/atmos/RESTART/20210801.000000.coupler.res
-rw-r----- 1 cmartin da   300 Apr 25  2022 /work2/noaa/da/cmartin/UFO_eval/data/para/output_ufo_eval_aug2021/2021080100/gdas.20210731/18/atmos/RESTART/20210801.010000.coupler.res
-rw-r----- 1 cmartin da   300 Apr 25  2022 /work2/noaa/da/cmartin/UFO_eval/data/para/output_ufo_eval_aug2021/2021080100/gdas.20210731/18/atmos/RESTART/20210801.020000.coupler.res
-rw-r----- 1 cmartin da   300 Apr 25  2022 /work2/noaa/da/cmartin/UFO_eval/data/para/output_ufo_eval_aug2021/2021080100/gdas.20210731/18/atmos/RESTART/20210801.030000.coupler.res
```
I'll chgrp to da

azadeh-gh commented:
@CoryMartin-NOAA Thank you!

CoryMartin-NOAA committed:
Merge branch 'develop' into feature/gdas-validation

RussTreadon-NOAA commented:
@azadeh-gh , you [mentioned]([float32_fill_value]) using 1.e11 for `solarAzimuthAngle`.  I modified my working copy of `bufr2ioda_sevcsr.py` to do this.    `fv3jedi_var.x` ran past the `MetaData/solarAzimuthAngle` error, but it still aborted.  The new traceback is
```
  9: Exception:         source_column:  0
  9:    source_filename:        /work2/noaa/da/rtreadon/gdas-validation/global-workflow/sorc/gdas.cd/ioda/src/engines/ioda/include/ioda/Variables/Variable.h
  9:    source_function:        Variable_Implementation ioda::detail::Variable_Base<Variable_Implementation>::read(gsl::span<DataType>, const ioda::Select\
ion &, const ioda::Selection &) const [with DataType = int; Marshaller = ioda::detail::Object_Accessor_Regular<int, int>; TypeWrapper = ioda::Types::GetType_Wrapper<int, 0>; Variable_Implementation = ioda::Variable]
  9:    source_line:    539
  9:
  9: Exception: oops::Variational<FV3JEDI, UFO and IODA observations> terminating...
```
I'm not confident that I added `solarAzimuthAngle` correctly.

While running gdasprepatmiodaobs I noticed the follow warning messages from `bufr2ioda_sevcsr.py`
```
Warning: Query String */SDTB/TMBRST didn't apply to subset NC021043
Warning: Query String */TMBRST didn't apply to subset NC021043
Warning: Query String */SCCF didn't apply to subset NC021043
Warning: Query String */NCLDMNT didn't apply to subset NC021043
Warning: Query String */SIID[1] didn't apply to subset NC021043
```
Are these warning messages expected?

emilyhcliu committed:
Minor observation YAML updates for ATMS, SATWIND and SCATWIND (#789)

This PR includes the following:
1. YAML name change - add sensor (ABI) to the filename of observation
YAML for SATWIND from GOES platforms
2. YAML name change - add sensor (ASCAT) to the filename of observation
YAML for SCATWIND from MetOp platforms
3. update the YAML name changes in the list for 3dvar 
4. Add `CO2` in obs forward operator and add `linear obs operator`
without CO2 for ATMS

These changes have no impact on 3DVar results.

CoryMartin-NOAA committed:
Merge branch 'develop' into feature/gdas-validation

emilyhcliu committed:
Update IASI  YAML files for end-to-end testing (#769)

RussTreadon-NOAA commented:
@azadeh-gh , is there any way in which I can help with troubleshooting?

azadeh-gh commented:
@RussTreadon-NOAA Thanks Russ, I added  solaziang=np.array([float32_fill_value], dtype=np.float32) for solarAzimuthAngle.
I'm running the JEDI 3D-Var, and found other parameters that need to be added.

RussTreadon-NOAA commented:
Thank you @azadeh-gh for systematically identifying and resolving issues with seviri processing.

RussTreadon-NOAA committed:
Merge branch 'develop' into feature/gdas-validation

RussTreadon-NOAA committed:
Merge branch 'develop' into feature/gdas-validation

emilyhcliu committed:
Update observation YAMLs for conventional surface pressure (#803)

This PR includes the following updates for conventional surface pressure
1. Add observation YAML under configuration
2. Update observation YAML under testing
3. Add conv_ps to 3dvar obs list
4. Set window shift to true (time boundary is inclusive)
5. Validate 3dvar for surface (station) pressure (end-to-end)
- [QC
validation](https://github.com/NOAA-EMC/GDASApp/pull/803#issuecomment-1851387894)
-
[Increments](https://github.com/NOAA-EMC/GDASApp/pull/803#issuecomment-1852496937)

See [PR #792 ](https://github.com/NOAA-EMC/GDASApp/pull/792) for the
BUFR converter work to convert the following surface pressure
observation into IODA:
- SFC - 181, 187
- SFCSHIP - 180, 183
- SONDE - 120

The related UFO evaluation (using GSI obs and geovals) results for
surface pressure are documented
[here](https://docs.google.com/presentation/d/1OvVMYOutxVyk0GE93Jzoh3hQVoj-dCgrwsVWDDlPT1s/edit?usp=sharing).

Commit:
Updated bufr2ioda_sevcsr.py (added new variables)and json file for converting SEVIRI METEOSAT-8 and METEOSAT-11 from BUFR to IODA. Updated the seviri_m08.yaml and seviri_m11.yaml in the config, added data thinning, filters for sensorZenithAngle, cloud fraction check, and WV Channels 5-6 scene consistency check.

emilyhcliu commented:
@azadeh-gh needs to fix coding norm errors

Commit:
Merged branch 'feature/gdas-validation' and fixed coding norm errors
Merge branch 'feature/gdas-validation' of https://github.com/NOAA-EMC/GDASApp into feature/Azadeh_gdas-validation

Conflicts:
	parm/atm/obs/lists/gdas_prototype_3d.yaml

Commit:
updated cp0.sh and gen_prototype.sh

Commit:
Merged "develop" branch
Merge branch 'develop' of https://github.com/NOAA-EMC/GDASApp into feature/Azadeh_gdas-validation

Conflicts:
	prototypes/configs/cp0.sh

emilyhcliu commented:
@azadeh-gh Thanks for bringing the branch updated with the develop!   I am about to run the test (ORION will be off tomorrow).

nicholasesposito commented:
I think this needs a new merge. It says there are 235 files changed.

emilyhcliu commented:
@azadeh-gh There are 235 files changed.  This is odd.  

> I think this needs a new merge. It says there are 235 files changed.

I just found the same thing. @azadeh-gh you changed 5 files, I remembered.

emilyhcliu commented:
> @azadeh-gh There are 235 files changed. This is odd.
> 
> > I think this needs a new merge. It says there are 235 files changed.
> 
> I just found the same thing. @azadeh-gh you changed 5 files, I remembered.

I changed the base from feature/gdas-validation to develop. 
Now, the changes are down from 235 to 5.

Commit:
Added  dim_list=["Location", "Channel"] to "MetaData/cloudFree" and "MetaData/cloudAmount"

Commit:
Updated the python converter to only read channel 5 CloudFree values from bufr.Modified the filters in the YAMLs

RussTreadon-NOAA commented:
Please update `feature/Azadeh_gdas-validation` with recent changes from GDASApp `develop`.  Note that file `parm/atm/obs/lists/gdas_prototype_3d.yaml` conflicts with `develop`.

RussTreadon-NOAA commented:
@azadeh-gh , [`feature/Azadeh_gdas-validation`](https://github.com/NOAA-EMC/GDASApp/tree/feature/Azadeh_gdas-validation) is 47 commits behind `develop`.  Updating with `develop` is straightforward with the following caveat.

File `parm/atm/obs/lists/gdas_prototype_3d.yaml` no longer exists in `develop`.  The replacement file is named `parm/atm/obs/lists/gdas_prototype_3d.yaml.j2`.  Your older copy of the file activates two observation yamls
```
- !INC ${OBS_YAML_DIR}/seviri_m08.yaml
- !INC ${OBS_YAML_DIR}/seviri_m11.yaml
```
These yamls are not listed in `gdas_prototype_3d.yaml.j2`.  Adding them is not as simple as adding the above serviri yaml files.  The above seviri yaml file need to be converted to jinja2 templates and then `gdas_prototype_3d.yaml.j2` updated to include the seviri jinja2 templates.

Do you want assistance with this?

azadeh-gh commented:
@RussTreadon-NOAA Do I need to update the seviri YAMLs to convert to jinja2?

RussTreadon-NOAA commented:
Yes, the seviri yamls need to be updated to jinja2 templates

- replace `$( foo )` with `{{ foo }}`
- add single quotes around filenames on `obsfile` lines
- add two blank spaces at the start of each line in the yaml file.  shifting comments is options.  The one exception to this is the first line.  `obs space` shifts over two spaces as `- obs space` 
- change file suffix from `.yaml` to `.yaml.j2`

Beware:  If you change the file suffix via `git mv` and in the same commit modify the _new_ jinja2 template, the previous history of the yaml is lost.   git treats the combination of these two actions as creation of a new file  I think two commits will preserve the history.  One commit modifies the file.  Another commit moves the file from _filename1_ to _filename2_.

Commit:
Added seviri_m08.yaml.j2 and seviri_m11.yaml.j2

Commit:
added gdas_prototype_3d.yaml.j2

Commit:
seviri_m08.yaml.j2 modifications

 Please enter the commit message for your changes. Lines starting

Commit:
Remove unmerged and unstaged files

Commit:
Modified YAMLS

azadeh-gh commented:
@RussTreadon-NOAA Thank you. I updated the YAMLS to jinja2.

emilyhcliu commented:
Reviewers.  Please look at the [T2O Issue #99](https://github.com/NOAA-EMC/JEDI-T2O/issues/99) related to this PR for test results.

@azadeh-gh  I have two comments:
1. There are many plots in the [issue #99](https://github.com/NOAA-EMC/JEDI-T2O/issues/99)
I think it would be good to have a short summary of your validation results here.  
2.  Please add error handling in the Python converter for missing input files or missing subsets
     - the need to add the error handling was identified by @RussTreadon-NOAA  (see [PR #937](https://github.com/NOAA-EMC/GDASApp/pull/937) for more infomation)
     - here is the example of error handling added in the converter
    
![Screenshot 2024-03-21 at 11 34 43 PM](https://github.com/NOAA-EMC/GDASApp/assets/36091766/52531e2e-818f-470c-9d95-b47af758d9d6)

RussTreadon-NOAA commented:
@azadeh-gh , the conflict in `parm/atm/obs/lists/gdas_prototype_3d.yaml.j2` is easily resolved.  `develop` adds Emily's ozone work to the obs list.  Your branch has seviri.  Both can ozone and seviri can be in the obs list.

Commit:
Added error handling to the python converter. Added Ozone to the obs lis to resolve the conflict in gdas_prototype_3d.yaml.j2

azadeh-gh committed:
Merge branch 'develop' into feature/Azadeh_gdas-validation

azadeh-gh commented:
![SEVIRI-QC-Flowchart pptx](https://github.com/NOAA-EMC/JEDI-T2O/assets/36091766/168931da-a9f3-4e3b-9708-75102492b1cf)

### Summary of validation results

**No Thinning without QC (UFO vs. GSI) - Channel 5 Time Level 3** 

![ufo_seviri_m08_omf_bc_channel5_time1](https://github.com/NOAA-EMC/JEDI-T2O/assets/36091766/3d669ce4-eac6-4769-8b08-11842cf6cd20)
![gsi_seviri_m08_omf_gsi_bc_channel5_time1](https://github.com/NOAA-EMC/JEDI-T2O/assets/36091766/f4cd2c92-6cf6-458a-a85d-6d089f7b284c)

**No Thinning with QC: UFO vs. GSI - Channel 5  Time Level 1**

![ufo_seviri_m08_omf_qc_channel5_time1](https://github.com/NOAA-EMC/JEDI-T2O/assets/36091766/4cf48859-f7e6-4510-be2e-f5b7599d25ca)
![gsi_seviri_m08_omf_gsi_qc_channel5_time1](https://github.com/NOAA-EMC/JEDI-T2O/assets/36091766/3c2b6f67-cbc9-405b-8c11-cfdaa3a92fda)

![GSI_OmF_Seviri08_ch5_Thinned_Passed_QCs](https://github.com/NOAA-EMC/JEDI-T2O/assets/51101867/a5d96210-8242-4947-ba16-089ec2ae7642)
![JEDI_OmF_Seviri08_CH5_Thinned_PassedQCs](https://github.com/NOAA-EMC/JEDI-T2O/assets/51101867/80a80c46-12e8-4458-98c6-d33a1b793e9d)

**Here are the channel statistics plots (#obs, bias/std for OMF and final obs error) for JEDI and GDAS runs (Thinned)**

**O-F Bias/Standard Deviation**
JEDI            |  GDAS (original: ptime=0 & ithin_time=1)
:-------------------------:|:-------------------------:
![](https://github.com/NOAA-EMC/JEDI-T2O/assets/36091766/c80322e3-b10e-4b4d-be7b-63bf40f93477) | ![](https://github.com/NOAA-EMC/JEDI-T2O/assets/36091766/67c06817-11a2-411c-ab27-c3492853eb92)

**Final Obs Error**
JEDI            |  GDAS (original: ptime=0 & ithin_time=1)
:-------------------------:|:-------------------------:
![](https://github.com/NOAA-EMC/JEDI-T2O/assets/36091766/1408e155-31ed-4376-99b7-3c497f8e57f1)|![](https://github.com/NOAA-EMC/JEDI-T2O/assets/36091766/a81df486-a45f-44b3-8173-ddbe21714065)

**Number Obs Passed QC**
JEDI            |  GDAS (original: ptime=0 & ithin_time=1)
:-------------------------:|:-------------------------:
![](https://github.com/NOAA-EMC/JEDI-T2O/assets/36091766/6f3c9f76-bcd2-4725-8b03-c8b91750cf10)|![](https://github.com/NOAA-EMC/JEDI-T2O/assets/36091766/6e1fcab9-621d-4d01-bb13-b0f99fd7bb85)

**O-F Bias/Standard Deviation**
JEDI            |  GDAS (ptime=0 & ithin_time=2)
:-------------------------:|:-------------------------:
![](https://github.com/NOAA-EMC/JEDI-T2O/assets/36091766/889db7ca-f242-4b47-a3df-2ee3f67707f2) | ![](https://github.com/NOAA-EMC/JEDI-T2O/assets/36091766/fcce39a9-03da-4f17-bcdd-8f5714860de2)

**Final Obs Error**
JEDI            |  GDAS (ptime=0 & ithin_time=2)
:-------------------------:|:-------------------------:
![](https://github.com/NOAA-EMC/JEDI-T2O/assets/36091766/5f99ec21-30a8-46b3-b7a6-36a6e571f320) | ![](https://github.com/NOAA-EMC/JEDI-T2O/assets/36091766/d9d8fba1-628a-4c8d-8a6b-a905466bb4e0)

**Number Obs Passed QC**
JEDI            |  GDAS (ptime=0 & ithin_time=2)
:-------------------------:|:-------------------------:
![](https://github.com/NOAA-EMC/JEDI-T2O/assets/36091766/3dd2a2fd-c766-4adc-a759-9ea2af1c0a7e)| ![](https://github.com/NOAA-EMC/JEDI-T2O/assets/36091766/b14387c9-57d5-46dd-812a-9a318084c88c)

azadeh-gh commented:
> @azadeh-gh , the conflict in `parm/atm/obs/lists/gdas_prototype_3d.yaml.j2` is easily resolved. `develop` adds Emily's ozone work to the obs list. Your branch has seviri. Both can ozone and seviri can be in the obs list.

Thank you Russ. Updated the `parm/atm/obs/lists/gdas_prototype_3d.yaml.j2` to have both SEVIRI and Ozone in the obs list.

azadeh-gh commented:
> Reviewers. Please look at the [T2O Issue #99](https://github.com/NOAA-EMC/JEDI-T2O/issues/99) related to this PR for test results.
> 
> @azadeh-gh I have two comments:
> 
> 1. There are many plots in the [issue #99](https://github.com/NOAA-EMC/JEDI-T2O/issues/99)
>    I think it would be good to have a short summary of your validation results here.
> 2. Please add error handling in the Python converter for missing input files or missing subsets
>    
>    * the need to add the error handling was identified by @RussTreadon-NOAA  (see [PR #937](https://github.com/NOAA-EMC/GDASApp/pull/937) for more infomation)
>    * here is the example of error handling added in the converter
> 
> ![Screenshot 2024-03-21 at 11 34 43 PM](https://private-user-images.githubusercontent.com/36091766/315877700-52531e2e-818f-470c-9d95-b47af758d9d6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE0ODA2NTIsIm5iZiI6MTcxMTQ4MDM1MiwicGF0aCI6Ii8zNjA5MTc2Ni8zMTU4Nzc3MDAtNTI1MzFlMmUtODE4Zi00NzBjLTlkOTUtYjQ3YWY3NThkOWQ2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzI2VDE5MTIzMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWVmOWIyZTFjMjM2NjAzYTQwN2JmMjRiOWU2NWRmMDEyZTZmZGFjMzM0YWMyMWRlMTE5ZTEwZjMxMmYzNjJkMDImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.x963PDN1v3Jn45FJoi0HTjs2SGZ8xwcR4mPrRO9vxpw)

Thank you for your suggestions. 
1. Added summary of our evaluation results from the [Issue #99](https://github.com/NOAA-EMC/JEDI-T2O/issues/99) here.
2. Updated my python converter and added error handling for missing input files or missing subsets.

RussTreadon-NOAA commented:
@azadeh-gh , I went to test `feature/Azadeh_gdas-validation` for a near real-time case (2024022400).  The `sevcsr` bufr dump file is zero length.  Your code noted this and properly exited. That's good.   A check of the gdas directory in the GDA shows all `gdas.t*z.sevcsr.tm00.bufr_d` files for 2024 are, thus far, zero length.

Do we know why all 2024 `sevcsr` dump files are zero length?  Will the data return?

RussTreadon-NOAA reviewed (commented):
Changes look reasonable.  Unable to test in JEDI ATM CI case.  Bufr dump file `gdas.t00z.sevcsr.tm00.bufr_d` is zero length.

```
-rw-r--r-- 1 wkolczyn global 0 Feb 24 05:53 /work/noaa/rstprod/dump/gdas.20240224/00/atmos/gdas.t00z.sevcsr.tm00.bufr_d
```

ADCollard commented:
@RussTreadon-NOAA We cannot do a real time test of SEVIRI as the CSR stream has been discontinued.   The same information is in the ASR stream, so the next step after this is to create an ASR converter.

ADCollard reviewed (approved):
The results in [JEDI-T2O Issue 99](https://github.com/NOAA-EMC/JEDI-T2O/issues/99) look good.  Approve.

RussTreadon-NOAA commented:
Thank you @ADCollard for explaining what I am seeing.
